pkg/pillar/cmd/nodeagent/handletimers.go

```diff
@@ -0,0 +1,188 @@
+func handleResetOnCloudDisconnect(ctx *nodeagentContext) {
+	// apply the reset time function
+	resetLimit := time.Second * time.Duration(globalConfig.ResetIfCloudGoneTime)
+	timePassed := time.Since(ctx.lastConfigReceivedTime)
+	if timePassed > resetLimit {
+		errStr := fmt.Sprintf("Exceeded outage for cloud connectivity %d by %d seconds; rebooting\n",
+			resetLimit/time.Second,
+			(timePassed-resetLimit)/time.Second)
+		log.Errorf(errStr)
+		execReboot(ctx, errStr)
+	}
+}
+
+// baseos upgrade test validation utilities
+func setTestStartTime(ctx *nodeagentContext) {
+	// test has already started or, the current
+	// partition is not in inprogress state,
+	// nothing to do
+	if !ctx.updateInprogress || ctx.testInprogress {
+		return
+	}
+	// only when current partition state is in progress
+	// start the test
+	ctx.upgradeTestStartTime = time.Now()
+	ctx.testInprogress = true
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	ctx.remainingTestTime = successLimit
+	ctx.testInprogress = true
+}
+
+func resetTestStartTime(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	// the led status has changed, incase, the test is in progress,
+	// we need to start the test again, reset first
+	ctx.testInprogress = false
+	ctx.remainingTestTime = 0
+	ctx.testInprogress = false
+}
+
+func setConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+}
+
+func resetConfigGetFailState(ctx *nodeagentContext) {
+	if ctx.configGetFail {
+		ctx.configGetFail = false
+		ctx.configGetFailCount = 0
+	}
+}
+
+func executeConfigGetFailState(ctx *nodeagentContext) {
+	if !ctx.configGetFail {
+		return
+	}
+	if ctx.configGetFailCount >= maxConfigGetFailCount {
+		resetConfigGetFailState(ctx)
+		resetTestStartTime(ctx)
+	}
+}
+
+func checkUpgradeValidationTestTimeExpiry(ctx *nodeagentContext) bool {
+	if !ctx.testInprogress {
+		return false
+	}
+	timePassed := time.Since(ctx.upgradeTestStartTime)
+	successLimit := time.Second *
+		time.Duration(globalConfig.MintimeUpdateSuccess)
+	if timePassed < successLimit {
+		ctx.remainingTestTime = successLimit - timePassed
+		log.Infof("CurPart: %s inprogress, waiting for %d seconds\n",
+			ctx.curPart, ctx.remainingTestTime/time.Second)
+		publishNodeAgentStatus(ctx)
+		return false
+	}
+	return true
+}
+
+func handleConfigGetFail(ctx *nodeagentContext) {
+	if !ctx.testInprogress {
+		return
+	}
+	if !ctx.configGetFail {
+		ctx.configGetFail = true
+	}
+	ctx.configGetFailCount++
+	executeConfigGetFailState(ctx)
+}
+
+func updateLastConfigReceivedTime(ctx *nodeagentContext,
+	status types.ZedAgentStatus) {
+	log.Debugf("lastConfigReceivedtime: %v, failCount: %d \n",
+		status.LastConfigReceivedTime, status.ConfigGetFailCount)
+	// time stamp has not changed, since last config receive
+	if ctx.lastConfigReceivedTime == status.LastConfigReceivedTime {
+		handleConfigGetFail(ctx)
```

Comment:
How is a lack of a change a failure? You don't know whether you might get false notifications from pubsub or whether some other field in the published data has changed.